zendesk/importer.go

```diff
@@ -0,0 +1,187 @@
+	ExternalID          interface{}   `json:"external_id"`
+	Via                 Via           `json:"via"`
+	CreatedAt           time.Time     `json:"created_at"`
+	UpdatedAt           time.Time     `json:"updated_at"`
+	Type                interface{}   `json:"type"`
+	Subject             string        `json:"subject"`
+	RawSubject          string        `json:"raw_subject"`
+	Description         string        `json:"description"`
+	Priority            interface{}   `json:"priority"`
+	Status              string        `json:"status"`
+	Recipient           interface{}   `json:"recipient"`
+	RequesterID         int64         `json:"requester_id"`
+	SubmitterID         int64         `json:"submitter_id"`
+	AssigneeID          int64         `json:"assignee_id"`
+	OrganizationID      int64         `json:"organization_id"`
+	CollaboratorIds     []interface{} `json:"collaborator_ids"`
+	FollowerIds         []interface{} `json:"follower_ids"`
+	EmailCcIds          []interface{} `json:"email_cc_ids"`
+	ForumTopicID        interface{}   `json:"forum_topic_id"`
+	ProblemID           interface{}   `json:"problem_id"`
+	HasIncidents        bool          `json:"has_incidents"`
+	IsPublic            bool          `json:"is_public"`
+	DueAt               interface{}   `json:"due_at"`
+	Tags                []interface{} `json:"tags"`
+	CustomFields        []interface{} `json:"custom_fields"`
+	SatisfactionRating  interface{}   `json:"satisfaction_rating"`
+	SharingAgreementIds []interface{} `json:"sharing_agreement_ids"`
+	Fields              []interface{} `json:"fields"`
+	FollowupIds         []interface{} `json:"followup_ids"`
+	TicketFormID        int64         `json:"ticket_form_id"`
+	AllowChannelback    bool          `json:"allow_channelback"`
+	AllowAttachments    bool          `json:"allow_attachments"`
+	GeneratedTimestamp  int           `json:"generated_timestamp"`
+}
+
+type Via struct {
+	Channel string `json:"channel"`
+	Source  Source `json:"source"`
+}
+
+type Source struct {
+	From interface{} `json:"from"`
+	To   interface{} `json:"to"`
+	Rel  interface{} `json:"rel"`
+}
+
+type BulkImporter struct {
+	userName   string       // userName to login as admin to zendesk
+	apiToken   string       // token to authenticate the user
+	client     *http.Client // http client to connect zendesk
+	maxRetries uint64       // max API retries in case of 429, before returning error
+	baseURL    string       // zendesk api url
+	retryCount uint64       // number of retry count made for current data
+}
+
+// NewBulkImporter initialize bulk importer to write bulk tickets to zendesk
+func NewBulkImporter(userName, apiToken, domain string, maxRetries uint64) *BulkImporter {
+	return &BulkImporter{
+		//url:        fmt.Sprintf("https://%s.zendesk.com/api/v2/imports/tickets/create_many", domain),
+		client:     newHTTPClient(),
+		userName:   userName,
+		apiToken:   apiToken,
+		baseURL:    fmt.Sprintf("https://%s.zendesk.com", domain),
```

Comment:
IIUC, it's possible to have custom URLs for ZenDesk instances, in which case this won't work. However, I think we can open an issue for that and handle it separately.

Reply:
AFAIK, the custom URLs are DNS CName-based redirects to the Zendesk subdomain and not self-hosted instances. I am not sure if we will need to handle custom URLs. 
https://support.zendesk.com/hc/en-us/articles/4408823719578-Does-Zendesk-offer-self-hosting-options-

Reply:
Gotcha, good point. :+1: